pkg/ocm/servicelog.go

```diff
@@ -0,0 +1,88 @@
+package ocm
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/openshift/ocm-agent-operator/api/v1alpha1"
+	"github.com/openshift/ocm-agent/pkg/consts"
+	"github.com/prometheus/alertmanager/template"
+)
+
+const (
+	ServiceLogActivePrefix  = "Issue Notification"
+	ServiceLogResolvePrefix = "Issue Resolution"
+)
+
+type ServiceLog struct {
+	ServiceName   string                               `json:"service_name"`
+	ClusterUUID   string                               `json:"cluster_uuid,omitempty"`
+	Summary       string                               `json:"summary"`
+	Description   string                               `json:"description"`
+	InternalOnly  bool                                 `json:"internal_only"`
+	Severity      v1alpha1.NotificationSeverity        `json:"severity"`
+	LogType       string                               `json:"log_type,omitempty"`
+	DocReferences []v1alpha1.NotificationReferenceType `json:"doc_references,omitempty"`
+}
+
+func NewServiceLog(summary, firingDesc, resolveDesc, clusterID string, severity v1alpha1.NotificationSeverity, logType string, references []v1alpha1.NotificationReferenceType, firing bool) *ServiceLog {
+	sl := ServiceLog{
+		ServiceName:   consts.ServiceLogServiceName,
+		ClusterUUID:   clusterID,
+		Summary:       summary,
+		InternalOnly:  false,
+		Severity:      severity,
+		LogType:       logType,
+		DocReferences: references,
+	}
+
+	// Use different Summary and Description for firing and resolved status for an alert
+	if firing {
+		sl.Description = firingDesc
+		sl.Summary = ServiceLogActivePrefix + ": " + summary
+	} else {
+		sl.Description = resolveDesc
+		sl.Summary = ServiceLogResolvePrefix + ": " + summary
+	}
+
+	return &sl
+}
+
+var (
+	slVarRefRe = regexp.MustCompile(`\${[^{}]*}`)
+)
+
+func (sl *ServiceLog) ReplacePlaceHoldersWithAlertLabelsAndAnnotations(alert *template.Alert) error {
+	replacePlaceHoldersInString := func(s string) (string, error) {
+		var err error
+		resolvePlaceHolder := func(placeHolder string) string {
+			if err == nil {
+				key, value, isOk := placeHolder[2:len(placeHolder)-1], "", false
+
+				if value, isOk = alert.Labels[key]; !isOk {
+					if value, isOk = alert.Annotations[key]; !isOk {
+						err = fmt.Errorf("alert has no '%s' label or annotation which could be used to replace place holders in the template", key)
+
+						return placeHolder
+					}
+				}
+				return value
+			}
+
+			return placeHolder
+		}
+
+		return slVarRefRe.ReplaceAllStringFunc(s, resolvePlaceHolder), err
+	}
+
+	var err error
+
+	if sl.Summary, err = replacePlaceHoldersInString(sl.Summary); err != nil {
+		return err
+	}
```

Comment:
Would save two lines? :) 

```suggestion
	if sl.Summary, err := replacePlaceHoldersInString(sl.Summary); err != nil {
		return err
	}
```

Reply:
I think that `:=` declare and define/set a variable...
... and `sl.Summary` cannot be declared as it is already declared by being a member of `sl`.

`=` does not have this problem.

Reply:
`:=` works here as well, as it works if only one of the variables needs a fresh initialization. It would only initialize `err` and replace `sl.Summary`

Reply:
`:=` only works for variables, not for members:
https://go.dev/play/p/5yJGO_eVEZ8

Reply:
TIL. Thanks! :)